lake/branch.go

```diff
@@ -90,252 +54,223 @@ func (b *Branch) Load(ctx context.Context, r zio.Reader, date nano.Ts, author, m
 		if err != nil {
-			return err
+			return nil, err
 		}
-		b.base, err = b.snapParentAt(ctx, basePtr)
-	}
-	return err
+		return object, nil
+	})
+}
+
+func (b *Branch) mergeInto(ctx context.Context, parent *Branch, author, message string) (ksuid.KSUID, error) {
+	if b == parent {
+		return ksuid.Nil, errors.New("cannot merge branch into itself")
+	}
+	return parent.commit(ctx, func(head *branches.Config, retries int) (*commits.Object, error) {
+		return b.buildMergeObject(ctx, head, retries, author, message, parent.Name)
+	})
+	//XXX we should follow parent commit with a child rebase... do this
+	// next... we want to fast forward the child to any pending commits
+	// that happened on the child branch while we were merging into the parent.
+	// and rebase the child branch to point at the parent where we grafted on
+	// it's ok if new commits are arriving past the parent graft on point...
 }
 
-// Snapshot returns a snapshot of this branch at the journal entry where
-// tag is a commit.  If tag is ksuid.Nil, the snapshot is at the tip of
-// the branch.
-func (b *Branch) Snapshot(ctx context.Context, tag ksuid.KSUID) (commit.View, error) {
-	if err := b.checkParent(ctx); err != nil {
+func (b *Branch) buildMergeObject(ctx context.Context, parent *branches.Config, retries int, author, message, parentName string) (*commits.Object, error) {
+	childPath, err := b.pool.commits.Path(ctx, b.Commit)
+	if err != nil {
 		return nil, err
 	}
-	if tag == ksuid.Nil {
-		// No commit reference.  Just play this branch's log
-		// on top of the parent.
-		if b.base == nil {
-			// No parent.
-			tip, err := b.log.Tip(ctx)
-			if err != nil {
-				if err != journal.ErrEmpty {
-					return nil, err
-				}
-				// Nothing in this branch.  Return parent snap.
-				return commit.NewSnapshot(), nil
-			}
-			return tip, nil
-		}
-		// We have a parent so create a patch and play this log
-		// into into the patch.
-		patch := commit.NewPatch(b.base)
-		if err := patch.PlayLog(ctx, b.log, 0); err != nil {
-			return nil, err
-		}
-		return patch, nil
+	parentPath, err := b.pool.commits.Path(ctx, parent.Commit)
+	if err != nil {
+		return nil, err
 	}
-	// We have a commit tag.  So find the journal ID (and branch in
-	// our ancesetry fort his tag.
-	branch, at, err := b.searchForTag(ctx, tag)
+	baseID := commonAncestor(parentPath, childPath)
+	if baseID == ksuid.Nil {
+		//XXX this shouldn't happen because because all of the branches
+		// should live in a single tree.
+		//XXX hmm, except if you branch main when it is empty...?
+		// we shoudl detect this and not allow it...?
+		return nil, errors.New("system error: cannot locate common ancestor for branch merge")
+	}
+	// Compute the snapshot of the common ancestor then compute patches
+	// along each branch and make sure the two patches do not have a
+	// delete conflict.  For now, this is the only kind of merge update
+	// conflict we detect.
+	base, err := b.pool.commits.Snapshot(ctx, baseID)
```

Comment:
What's the point of getting the base Snapshot? Should you just be able to just collect deletes from the two tips to the baseID and make sure there's no conflict between the two?

Reply:
You need to compare paths to detect write conflicts.  The tip is insufficient.  Right now, it's just detecting simple delete conflicts but this code path will get more sophisticated with fancier update models.